src/main/java/org/xbill/DNS/DNSSession.java

```diff
@@ -0,0 +1,129 @@
+    Message query = Message.newQuery(question);
+
+    return resolver.sendAsync(query).thenCompose(this::resolveRedirects);
+  }
+
+  private CompletionStage<LookupResult> resolveRedirects(Message response) {
+    CompletableFuture<LookupResult> future = new CompletableFuture<>();
+    maybeFollowRedirect(response, 1, future);
+    return future;
+  }
+
+  private void maybeFollowRedirect(
+      Message response, int redirectCount, CompletableFuture<LookupResult> future) {
+    try {
+      if (redirectCount > maxRedirects) {
+        throw new RedirectOverflowException(
+            format("Refusing to follow more than %s redirects", maxRedirects));
+      }
+
+      List<Record> allAnswers = response.getSection(Section.ANSWER);
+      LookupResult result = dealWithEmptyAnswer(allAnswers, response.getHeader().getRcode());
+      if (result != null) {
+        future.complete(result);
+        return;
+      }
+      Record firstAnswer = allAnswers.get(0);
+
+      if (firstAnswer.getType() == Type.DNAME || firstAnswer.getType() == Type.CNAME) {
+        resolver
+            .sendAsync(Message.newQuery(buildRedirectQuery(response)))
+            .thenAccept(m -> maybeFollowRedirect(m, redirectCount + 1, future));
+      } else {
+        future.complete(new LookupResult(allAnswers, null));
+      }
+    } catch (LookupFailedException e) {
+      future.completeExceptionally(e);
+    }
+  }
+
+  private Record buildRedirectQuery(Message response) {
+    List<Record> answer = response.getSection(Section.ANSWER);
+    Record firstAnswer = answer.get(0);
+    if (answer.size() != 1) {
+      throw new InvalidZoneDataException("Multiple CNAME RRs not allowed, SEE RFC1034 3.6.2");
+    }
+    Record question = response.getQuestion();
+    if (firstAnswer.getType() == Type.CNAME) {
+      return Record.newRecord(
+          ((CNAMERecord) firstAnswer).getTarget(), question.getType(), question.getDClass());
+    }
+    // It's not a CNAME, it's a DNAME
+    try {
+      Name name = question.getName().fromDNAME((DNAMERecord) firstAnswer);
+      return Record.newRecord(name, question.getType(), question.getDClass());
+    } catch (NameTooLongException e) {
+      throw new InvalidZoneDataException(
+          "DNAME redirect would result in a name that would be too long");
+    }
+  }
+
+  /** Returns a LookupResult if this response was a non-exceptional empty result, else null. */
+  private LookupResult dealWithEmptyAnswer(List<Record> answer, int rcode) {
+    if (answer.isEmpty()) {
+      switch (rcode) {
+        case Rcode.NXDOMAIN:
+          return new LookupResult(answer, AdditionalDetail.NXDOMAIN);
+        case Rcode.NXRRSET:
+          return new LookupResult(answer, AdditionalDetail.NXRRSET);
+        case Rcode.SERVFAIL:
+          throw new ServerFailedException();
+        default:
+          throw new LookupFailedException(format("Unknown non-success error code %d", rcode));
```

Comment:
Use `Rcode.string(rcode)` to translate the numeric value. The underlying mnemonic class takes care of handling unknown values, so it's guaranteed to have a sensible message.

Reply:
That is great. At some point I think we should discuss moving to proper enums for the mnemonic classes, but that is for a later time.

Reply:
Enums: yes and no. A lot of stuff probably should stay as it is, not just because of the API compatibility, but also because DNS is extensible and with the integers everything still works when new stuff is introduced even if we don't update anything.

Reply:
Good point!